lib/master/master.go

```diff
@@ -0,0 +1,460 @@
+package master
+
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"math"
+	"math/rand"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"sort"
+	"time"
+
+	"github.com/aaronang/cong-the-ripper/lib"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+)
+
+type Master struct {
+	port             string
+	svc              *ec2.EC2 // safe to be used concurrently
+	instances        map[string]slave
+	jobs             map[int]*job
+	jobsChan         chan lib.Job
+	heartbeatChan    chan lib.Heartbeat
+	statusChan       chan chan statusSummary
+	newTasks         []*lib.Task
+	scheduledTasks   []*lib.Task
+	scheduleTicker   *time.Ticker // channel to instruct the main loop to schedule tasks
+	controllerTicker *time.Ticker
+	controller       controller
+	taskSize         int
+	quit             chan bool
+}
+
+type slave struct {
+	tasks    []*lib.Task
+	maxSlots int
+	instance *ec2.Instance // can't populate this easily
+}
+
+type controller struct {
+	dt       time.Duration
+	kp       float64
+	kd       float64
+	ki       float64
+	prevErr  float64
+	integral float64
+}
+
+type statusSummary struct {
+	instances map[string]slave // TODO *ec2.Instance probably won't serialise
+	jobs      map[int64]*job
+}
+
+// Init creates the master object
+func Init(port string) Master {
+	// set some defaults
+	return Master{
+		port:             port,
+		svc:              nil, // initialised in Run
+		instances:        make(map[string]slave),
+		jobs:             make(map[int]*job),
+		jobsChan:         make(chan lib.Job),
+		heartbeatChan:    make(chan lib.Heartbeat),
+		statusChan:       make(chan chan statusSummary),
+		newTasks:         make([]*lib.Task, 0),
+		scheduledTasks:   make([]*lib.Task, 0),
+		scheduleTicker:   nil, // initialised in Run
+		controllerTicker: nil, //initialised in Run
+		controller: controller{
+			dt:       time.Minute * 2,
+			kp:       1,
+			kd:       0,
+			ki:       0,
+			prevErr:  0,
+			integral: 0,
+		},
+		taskSize: 6400 * 1000 * 1000,
```

Comment:
Did you already benchmark how long it takes for an instance to hash this amount?